File: typescript/packages/core/src/client/x402Client.ts
/**
 * X402Client - x402 协议的核心支付客户端
 * 
 * 管理支付机制注册表并协调支付流程。
 */

import type {
  PaymentRequirements,
  PaymentPayload,
  PaymentPermitContext,
} from '../types/index.js';

/** 客户端机制接口 */
export interface ClientMechanism {
  /** 获取支付方案名称 */
  scheme(): string;
  
  /** 创建支付载荷 */
  createPaymentPayload(
    requirements: PaymentRequirements,
    resource: string,
    extensions?: { paymentPermitContext?: PaymentPermitContext }
  ): Promise<PaymentPayload>;
}

/** 客户端签名器接口 */
export interface ClientSigner {
  /** 获取签名器地址 */
  getAddress(): string;
  
  /** 签名原始消息 */
  signMessage(message: Uint8Array): Promise<string>;
  
  /** 签名类型化数据 (EIP-712) */
  signTypedData(
    domain: Record<string, unknown>,
    types: Record<string, unknown>,
    message: Record<string, unknown>
  ): Promise<string>;
  
  /** 检查代币授权 */
  checkAllowance(token: string, amount: bigint, network: string): Promise<bigint>;
  
  /** 确保足够的授权 */
  ensureAllowance(
    token: string,
    amount: bigint,
    network: string,
    mode?: 'auto' | 'interactive' | 'skip'
  ): Promise<boolean>;
}

/** 支付要求选择器函数 */
export type PaymentRequirementsSelector = (
  requirements: PaymentRequirements[]
) => PaymentRequirements;

/** 选择支付要求的过滤选项 */
export interface PaymentRequirementsFilter {
  scheme?: string;
  network?: string;
  maxAmount?: string;
}

/** 已注册的机制条目 */
interface MechanismEntry {
  pattern: string;
  mechanism: ClientMechanism;
  priority: number;
}

/**
 * X402Client - 核心支付客户端
 * 
 * 管理支付机制并协调支付流程。
 */
export class X402Client {
  private mechanisms: MechanismEntry[] = [];

  /**
   * 为网络模式注册支付机制
   * 
   * @param networkPattern - 网络模式（例如 "eip155:*", "tron:shasta"）
   * @param mechanism - 支付机制实例
   * @returns this 以支持链式调用
   */
  register(networkPattern: string, mechanism: ClientMechanism): X402Client {
    const priority = this.calculatePriority(networkPattern);
    this.mechanisms.push({
      pattern: networkPattern,
      mechanism,
      priority,
    });
    this.mechanisms.sort((a, b) => b.priority - a.priority);
    return this;
  }

  /**
   * 从可用选项中选择支付要求
   * 
   * @param accepts - 可用的支付要求
   * @param filters - 可选过滤器
   * @returns 选定的支付要求
   */
  selectPaymentRequirements(
    accepts: PaymentRequirements[],
    filters?: PaymentRequirementsFilter
  ): PaymentRequirements {
    let candidates = accepts;

    if (filters?.scheme) {
      candidates = candidates.filter(r => r.scheme === filters.scheme);
    }

    if (filters?.network) {
      candidates = candidates.filter(r => r.network === filters.network);
    }

    if (filters?.maxAmount) {
      const max = BigInt(filters.maxAmount);
      candidates = candidates.filter(r => BigInt(r.amount) <= max);
    }

    candidates = candidates.filter(r => this.findMechanism(r.network) !== null);

    if (candidates.length === 0) {
      throw new Error('No supported payment requirements found');
    }

    return candidates[0];
  }

  /**
   * 为给定要求创建支付载荷
   * 
   * @param requirements - 选定的支付要求
   * @param resource - 资源 URL
   * @param extensions - 可选扩展
   * @returns 支付载荷
   */
  async createPaymentPayload(
    requirements: PaymentRequirements,
    resource: string,
    extensions?: { paymentPermitContext?: PaymentPermitContext }
  ): Promise<PaymentPayload> {
    const mechanism = this.findMechanism(requirements.network);
    if (!mechanism) {
      throw new Error(`No mechanism registered for network: ${requirements.network}`);
    }

    return mechanism.createPaymentPayload(requirements, resource, extensions);
  }

  /**
   * 处理需要支付的响应
   * 
   * @param accepts - 可用的支付要求
   * @param resource - 资源 URL
   * @param extensions - 可选扩展
   * @param selector - 可选自定义选择器
   * @returns 支付载荷
   */
  async handlePayment(
    accepts: PaymentRequirements[],
    resource: string,
    extensions?: { paymentPermitContext?: PaymentPermitContext },
    selector?: PaymentRequirementsSelector
  ): Promise<PaymentPayload> {
    const requirements = selector
      ? selector(accepts)
      : this.selectPaymentRequirements(accepts);

    return this.createPaymentPayload(requirements, resource, extensions);
  }

  /**
   * 查找网络的机制
   */
  private findMechanism(network: string): ClientMechanism | null {
    for (const entry of this.mechanisms) {
      if (this.matchPattern(entry.pattern, network)) {
        return entry.mechanism;
      }
    }
    return null;
  }

  /**
   * 将网络与模式匹配
   */
  private matchPattern(pattern: string, network: string): boolean {
    if (pattern === network) return true;
    if (pattern.endsWith(':*')) {
      const prefix = pattern.slice(0, -1);
      return network.startsWith(prefix);
    }
    return false;
  }

  /**
   * 计算模式的优先级（更具体 = 更高优先级）
   */
  private calculatePriority(pattern: string): number {
    if (pattern.endsWith(':*')) return 1;
    return 10;
  }
}

File: python/x402/src/x402_tron/clients/x402_client.py
"""
X402Client - Core payment client for x402 protocol
"""

import logging
from typing import TYPE_CHECKING, Any, Callable, Protocol

from x402_tron.types import (
    PaymentPayload,
    PaymentRequirements,
)

if TYPE_CHECKING:
    from x402_tron.clients.token_selection import TokenSelectionStrategy

logger = logging.getLogger(__name__)


class ClientMechanism(Protocol):
    """Client mechanism interface"""

    def scheme(self) -> str:
        """Get the payment scheme name"""
        ...

    async def create_payment_payload(
        self,
        requirements: PaymentRequirements,
        resource: str,
        extensions: dict[str, Any] | None = None,
    ) -> PaymentPayload:
        """Create payment payload"""
        ...


PaymentRequirementsSelector = Callable[[list[PaymentRequirements]], PaymentRequirements]


class PaymentRequirementsFilter:
    """Filter options for selecting payment requirements"""

    def __init__(
        self,
        scheme: str | None = None,
        network: str | None = None,
    ):
        self.scheme = scheme
        self.network = network


class MechanismEntry:
    """Registered mechanism entry"""

    def __init__(self, pattern: str, mechanism: ClientMechanism, priority: int):
        self.pattern = pattern
        self.mechanism = mechanism
        self.priority = priority


class X402Client:
    """
    Core payment client for x402 protocol.

    Manages payment mechanism registry and coordinates payment flow.
    """

    def __init__(
        self,
        token_strategy: "TokenSelectionStrategy | None" = None,
    ) -> None:
        """
        Initialize X402Client.

        Args:
            token_strategy: Strategy for selecting which token to pay with.
                            If None, uses first available option.
        """
        self._mechanisms: list[MechanismEntry] = []
        self._token_strategy = token_strategy

    def register(self, network_pattern: str, mechanism: ClientMechanism) -> "X402Client":
        """
        Register a payment mechanism for a network pattern.

        Args:
            network_pattern: Network pattern (e.g., "eip155:*", "tron:shasta")
            mechanism: Payment mechanism instance

        Returns:
            self for method chaining
        """
        priority = self._calculate_priority(network_pattern)
        logger.info(
            f"Registering mechanism for pattern '{network_pattern}' with priority {priority}"
        )
        self._mechanisms.append(MechanismEntry(network_pattern, mechanism, priority))
        self._mechanisms.sort(key=lambda e: e.priority, reverse=True)
        return self

    async def select_payment_requirements(
        self,
        accepts: list[PaymentRequirements],
        filters: PaymentRequirementsFilter | None = None,
    ) -> PaymentRequirements:
        """
        Select payment requirements from available options.

        Applies filters, then delegates to the token selection strategy
        if one is configured (and a signer is available). Otherwise
        falls back to picking the first candidate.

        Args:
            accepts: Available payment requirements
            filters: Optional filters

        Returns:
            Selected payment requirements

        Raises:
            ValueError: No supported payment requirements found
        """
        logger.info(f"Selecting payment requirements from {len(accepts)} options")
        logger.debug(f"Available payment requirements: {[r.model_dump() for r in accepts]}")

        candidates = list(accepts)

        if filters:
            if hasattr(filters, "scheme") and filters.scheme:
                candidates = [r for r in candidates if r.scheme == filters.scheme]
                logger.debug(f"After scheme filter: {len(candidates)} candidates")
            if hasattr(filters, "network") and filters.network:
                candidates = [r for r in candidates if r.network == filters.network]
                logger.debug(f"After network filter: {len(candidates)} candidates")

        candidates = [r for r in candidates if self._find_mechanism(r.network) is not None]
        logger.debug(f"After mechanism filter: {len(candidates)} candidates")

        if not candidates:
            logger.error("No supported payment requirements found")
            raise ValueError("No supported payment requirements found")

        if self._token_strategy:
            selected = await self._token_strategy.select(candidates)
        else:
            selected = candidates[0]

        logger.info(
            "Selected payment requirement: network=%s, scheme=%s, amount=%s",
            selected.network,
            selected.scheme,
            selected.amount,
        )
        return selected

    async def create_payment_payload(
        self,
        requirements: PaymentRequirements,
        resource: str,
        extensions: dict[str, Any] | None = None,
    ) -> PaymentPayload:
        """
        Create payment payload for given requirements.

        Args:
            requirements: Selected payment requirements
            resource: Resource URL
            extensions: Optional extensions

        Returns:
            Payment payload
        """
        logger.info(
            f"Creating payment payload for network={requirements.network}, resource={resource}"
        )
        mechanism = self._find_mechanism(requirements.network)
        if mechanism is None:
            logger.error(f"No mechanism registered for network: {requirements.network}")
            raise ValueError(f"No mechanism registered for network: {requirements.network}")

        logger.debug(f"Using mechanism: {mechanism.__class__.__name__}")
        payload = await mechanism.create_payment_payload(requirements, resource, extensions)
        logger.info("Payment payload created successfully")
        return payload

    async def handle_payment(
        self,
        accepts: list[PaymentRequirements],
        resource: str,
        extensions: dict[str, Any] | None = None,
        selector: PaymentRequirementsSelector | None = None,
    ) -> PaymentPayload:
        """
        Handle payment required response.

        Args:
            accepts: Available payment requirements
            resource: Resource URL
            extensions: Optional extensions
            selector: Optional custom selector

        Returns:
            Payment payload
        """
        if selector:
            requirements = selector(accepts)
        else:
            requirements = await self.select_payment_requirements(accepts)

        return await self.create_payment_payload(requirements, resource, extensions)

    def _find_mechanism(self, network: str) -> ClientMechanism | None:
        """Find mechanism for network"""
        for entry in self._mechanisms:
            if self._match_pattern(entry.pattern, network):
                return entry.mechanism
        return None

    def _match_pattern(self, pattern: str, network: str) -> bool:
        """Match network against pattern"""
        if pattern == network:
            return True
        if pattern.endswith(":*"):
            prefix = pattern[:-1]
            return network.startswith(prefix)
        return False

    def _calculate_priority(self, pattern: str) -> int:
        """Calculate priority for pattern (more specific = higher priority)"""
        if pattern.endswith(":*"):
            return 1
        return 10
